### ISSUE TITLE
Py2 and 3 compatibility

### ISSUE THREAD

**Issue by nh2d** [imgix/imgix-python#40]
At Rover we are upgrading our codebase to python3.6 while maintaining compatibility with python2.7.

We are running into a problem where imgix isn't compatible with _both_ python 2.7 and 3.6 **at the same time.**

The reason is that the behavior of string literals in the imgix-python library is not consistent between python versions.

For example, take `URLBuilder.create_url(path, opts)`. If `path` is explicitly unicode in python 2, this will result in the following error: 

```
  File "/usr/local/lib/python2.7/dist-packages/imgix/urlhelper.py", line 188, in __str__
    "", ])
  File "/usr/local/lib/python2.7/dist-packages/future/backports/urllib/parse.py", line 387, in urlunparse
    _coerce_args(*components))
  File "/usr/local/lib/python2.7/dist-packages/future/backports/urllib/parse.py", line 115, in _coerce_args
    raise TypeError("Cannot mix str and non-str arguments")
TypeError: Cannot mix str and non-str arguments
```

This will fail in the same way if `path` is explicitly a byte string in python 3. The only way that it can be expected to work at all between the two versions is if the `path` variable is a string literal with no explicit type and [`from __future__ import unicode_literals`](https://python-future.org/compatible_idioms.html#unicode-text-string-literals) is not present.

In order to ease our transition of hundreds of thousands of lines of code to python3 we _need_ to include all future imports in all our files which means that `path` will be explicitly unicode in both python2 and python3.

The issue can be fixed by making string literals be either explicitly unicode or bytestring regardless of what version of python was being run. This can be done by adding `from __future__ import unicode_literals` to all the files in this library.

One way to automatically do this would be to use [futurize](https://python-future.org/futurize.html)

**Comment by sherwinski** [comment]
Hey @nh2d thanks for the ticket and sorry to hear that inconsistent string behavior is causing issues. If I am understanding correctly, it sounds like including all future imports is causing strings to be explicitly typed as unicode, which is causing the error. In trying to understand this issue I attempted to replicate the problem you mentioned without success. The steps I followed, per your description, were:

```python
Python 2.7.10
>>> from __future__ import unicode_literals
>>> import imgix
>>> domain = 'sherwinski.imgix.net'
>>> type(domain)
<type 'unicode'>
>>> imgix.UrlBuilder(domain).create_url('image.jpg')
'https://sherwinski.imgix.net/image.jpg?ixlib=python-2.0.0'
```

Perhaps it would help with my understanding if you could provide a more concrete example of how to replicate this. Also, I'd be happy to review a PR if you have the time and want to take a swing at the solution for this. Thanks!

**Comment by sectioneight** [comment]
I believe I tracked down the source of the issue, and it is not the fault of this library.

We have been using `futurize` to automatically convert our code, and one of the things it tries to do is monkey-patch the standard library with a call to `install_aliases` when using certain modules (e.g. `urllib`). With this patch activated, I can reproduce this error:

```python
>>> from future.standard_library import install_aliases
>>> install_aliases()
>>> from imgix import UrlBuilder
>>> UrlBuilder('anything').create_url('anything else')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/imgix/urlbuilder.py", line 112, in create_url
    return str(url_obj)
  File "/usr/local/lib/python2.7/dist-packages/imgix/urlhelper.py", line 188, in __str__
    "", ])
  File "/usr/local/lib/python2.7/dist-packages/future/backports/urllib/parse.py", line 387, in urlunparse
    _coerce_args(*components))
  File "/usr/local/lib/python2.7/dist-packages/future/backports/urllib/parse.py", line 115, in _coerce_args
    raise TypeError("Cannot mix str and non-str arguments")
TypeError: Cannot mix str and non-str arguments
```

I don't think any change needs to be made to this library.

**Comment by jnak** [comment]
@ascandella Can you elaborate on how you think we should solve this issue in codebases that need to support both Python 2 and Python3?

As far as I understand, the issue it that in Python 3, `parse.urlunsplit` raises an error if the params mix both `unicode` and `string` (see https://github.com/PythonCharmers/python-future/issues/273#issuecomment-490569408). That means that codebases that target both Python 2 and Python 3 need to comply to that stricter behavior.

Currently, the strings created by `imgix` when running in Python 2/3 compat mode (ie using `future. standard_library.install_aliases`) are a mix of unicode and regular strings. For example, `self._scheme` is a regular strings while `query` is a unicode (because `urllib.quote` returns unicode).

The only solution that I see to this problem is to make sure that all strings created by `imgix` are unicode by annotating them with a leading u`...` prefix (see [explanation](https://python-future.org/compatible_idioms.html?highlight=unicode#unicode-text-string-literals)). This can be simplified to `from future import unicode_literals` but there may be [drawbacks](https://python-future.org/unicode_literals.html#drawbacks) That said I don't anticipate issues in `imgix` because it only manipulates URLs, which are usually treated as unicode anyway. This is the route that the Django has followed to support both Python 2 and 3.

The issue is actually larger than Python 2 / 3 compatibility since there are instrumentation libraries that call `install_aliases` under the hood such as [`opentracing_instrumenation`](https://github.com/uber-common/opentracing-python-instrumentation/search?q=install_aliases).

Would love to hear your thoughts on that. If you are onboard with this change, I'm happy to send a PR since this is currently blocking us.

**Comment by ascandella** [comment]
I think your evaluation is correct. If this library is going to be compatible with codebases that have called `install_aliases()` from future, then all of the strings passed to `urlunparse` need to be unicode (or non-unicode).

I haven't tested it, but it may be as simple as modifying https://github.com/imgix/imgix-python/blob/5d49f63e85cec9e3537700f1b1550dd2ebb497bc/imgix/urlhelper.py#L48 to add a `u`-prefix.

**Comment by ascandella** [comment]
> As far as I understand, the issue it that in Python 3, parse.urlunsplit raises an error if the params mix both unicode and string (see PythonCharmers/python-future#273 (comment)). That means that codebases that target both Python 2 and Python 3 need to comply to that stricter behavior.

Also just to clarify, the issue in Python 3 isn't mixing of `unicode` and `string` (since those are the same type). It's mixing of `bytes` and `str`

This is a Python 3 shell with just the standard library, no `future` / `install_aliases()`:

```
>>> from urllib.parse import urlunparse
>>> urlunparse([u'https', 'google.com', 'path', '', '', ''])
'https://google.com/path'
>>> urlunparse([b'https', 'google.com', 'path', '', '', ''])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/urllib/parse.py", line 456, in urlunparse
    _coerce_args(*components))
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/urllib/parse.py", line 120, in _coerce_args
    raise TypeError("Cannot mix str and non-str arguments")
TypeError: Cannot mix str and non-str arguments
```

**Comment by jnak** [comment]
@nh2d Can we re-open this issue? There are a simple things we can do to allow compatibility with Py2 and Py3. Would you consider a PR if I sent one?

**Comment by sherwinski** [comment]
Hi @jnak 👋 actually I am this project's maintainer. I'd be happy to look at a PR if you were to open one.

**Comment by jnak** [comment]
@sherwinski great! I just sent a [PR](https://github.com/imgix/imgix-python/pull/57) that fixes the issues and adds tests to prevent future regressions. I also verified that my branch fixes the original bug we found. Let me know if you have any questions.

**Comment by jnak** [comment]
@sherwinski When are you planning to release this patch?

**Comment by sherwinski** [comment]
Hey @jnak, this patch is now available in [v3.1.2](https://github.com/imgix/imgix-python/releases/tag/3.1.2) 🎉 
Thanks again for your help!
